src/python/pants/init/extension_loader.py

```diff
@@ -136,22 +136,38 @@ def load_backend(build_configuration: BuildConfiguration.Builder, backend_packag
     :raises: :class:``pants.base.exceptions.BuildConfigurationError`` if there is a problem loading
       the build configuration.
     """
-    backend_module = backend_package + ".register"
+
+    if templating_args:
+        kwargs = {'backend_package': backend_package}
+        kwargs.update(templating_args)
+        backend_module = kwargs.pop('template') + ".register"
```

Comment:
I think keeping this config key as `backend_package` would be more intuitive..
```suggestion
        backend_module = kwargs.pop('backend_package') + ".register"
```

as a non-POC, we'd of course want proper error handling to provide a helpful message in case of misconfiguration.

Reply:
I'm worried about the two uses of the word of `backend_package` which will then be in pants.toml. On the one hand we have `GLOBAL.backend_packages` which is presumably a list of backend packages. But if we leave the name here in `GLOBAL.templated_backends.<backend>.backend_package` as `backend_package` it would suggest it is usable as a `GLOBAL.backend_packages`.

Previous to this change, the members of `GLOBAL.backend_packages` were two things:
- a python package satisfying a particular interface that pants understands how to import and run.
- a set of pants Targets, Rules, build file aliases, etc: a pants _backend_ if you will. Pants creates the backend using the python package supplied above.

With templating, these two no longer map 1-1. The proposal so far in this POC is to map `GLOBAL.backend_packages` 1-1 with pants backends:

- `backend_package`: a "package" of instantiated rules, target definitions, etc.. Not necessarily mapping 1-1 with a python package, so it can include an alias to a templated backend.

Then we need to name the attribute referring to the python package that generates a templated backend. We could use something like `backend_template_package` or `template_package` instead of the bare `template` to make it clearer that the argument should refer to a python package.

Speaking in type pseudocode a bit more formally:
```python
GLOBAL.backend_packages: List[BackendPackage]

BackendPackage = RegularBackendPythonPackage | BackendPackageAlias

GLOBAL.templated_backends: Dict[BackendPackageAlias, TemplatedBackend]

class TemplatedBackend(TypedDict):
    template_package: TemplatePythonPackage
```

If instead we did:
```python
class TemplatedBackend(TypedDict):
    backend_package: TemplatePythonPackage
```
we would have the same attribute name referring to entities of different types.

Reply:
>  But if we leave the name here in GLOBAL.templated_backends.<backend>.backend_package as backend_package it would suggest it is usable as a GLOBAL.backend_packages.

Which it very well _could be_. I think I see these two as the same thing, in the same way I see two methods with different signatures still being methods.

The new thing we have, is that backends now may accept `kwargs` when being loaded (and with that we also support loading the same backend multiple times with different args). So to me, the template backend is still a backend, only that it expects additional arguments to be useful (typically).

Reply:
ah, just to note, we can still refer to the as template backend etc in the code.. my main concern is the user facing UX.

Reply:
I have one last option to consider, and just pushed it. Trying out `package`:
```toml
[GLOBAL.templated_backends.flake8]
package = "pants.backend.experimental.adhoc.code_quality_tool_backend_template"
goal = "lint"
target = "//:flake8_linter"
name = "Flake8"
```

With this we would still have no overloading. We can use "package" to refer to some python code that can produce rules, target definitions, etc. and "backend package" as the specialized version that does not require any more information to do so. The analogy to callables would be: backend package is to package as zero-arg callable is to callable.

What do you think?

Reply:
After the significant reworking I just did the discussion changes a lot. The toml for templated_backends now refers to a module and there is no `register.py` involved for these. Since it is no longer a python package, switched to referring to it back to `template`. Resolving the conversation for now, because the names probably need a completely fresh look given how different the backend generation API is in the current proposal.